### PR TITLE
Add Addon CRD e2e tests exclusively for openshift

### PR DESCRIPTION
**Pull request by nikhil-thomas**
# Changes

Add create and delete e2e tests for Addo CRD and reconciler

Modify test/e2e directory

Modify test/e2e-tests.sh

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 MODULE   = $(shell env GO111MODULE=on $(GO) list -m)
 DATE         ?= $(shell date +%FT%T%z)
 KO_DATA_PATH  = $(shell pwd)/cmd/$(TARGET)/kodata
-TARGET        = kubernetes
+TARGET        = ${E2E_TEST_PLATFORM}
```
Comment (vdemeester) [inline Makefile:4]: This changes will break `make apply` though.

File: test/e2e-common.sh [hunk 2]
```diff
@@ -27,6 +27,8 @@ source $(dirname $0)/../vendor/github.com/tektoncd/plumbing/scripts/e2e-tests.sh
 
 function install_operator_resources() {
   echo ">> Deploying Tekton Operator Resources"
+
+  E2E_TEST_PLATFORM=${E2E_TEST_PLATFORM:-kubernetes} \
   make apply || fail_test "Tekton Operator installation failed"
```
Comment (vdemeester) [inline test/e2e-common.sh:32]: `make TARGET=E2E_TEST_PLATFORM apply` and you are good to go, no need for changes in the `Makefile`

**Review by vdemeester — CHANGES_REQUESTED** [review]

File: docs/README.md [hunk 3]
```diff
@@ -19,8 +18,21 @@ If the files in `pkg/apis` are updated we need to run `codegen` scripts
...
+### Run operator
+
+**Target: Kubernetes**
+```shell script
+    make apply
+```
+
+**Target Openshift**
+```shell script
+    TARGET=openshift make apply
```
Comment (vdemeester) [inline docs/README.md:33]: or `make TARGET=openshift apply`, both work I think :stuck_out_tongue_closed_eyes:

File: test/README.md [hunk 4]
```diff
@@ -17,3 +17,22 @@ Then run:
 ```shell script
 ./test/e2e-tests.sh
 ```
+## Running TESTS
+
+if a cluster is already setup, then set `E2E_DEBUG` variable to skip initialization of a GKE cluster.
+
+```shell script
+export E@E_DEBUG=on
```
Comment (vdemeester) [inline test/README.md:25]: `s/E@E/E2E`

Comment (nikhil-thomas) [reply test/README.md:25]: :stuck_out_tongue:

Comment (nikhil-thomas) [reply test/README.md]: done :heavy_check_mark:

File: docs/README.md (hunk 3 above)
Comment (nikhil-thomas) [reply docs/README.md]: :heavy_check_mark: for uniformity

**Review by vdemeester — APPROVED** [review]
/cc @savitaashture

File: docs/README.md [hunk 5]
```diff
@@ -19,8 +18,21 @@ If the files in `pkg/apis` are updated we need to run `codegen` scripts
...
+**Target Openshift**
+```shell script
+    make TARGET=openshift apply
+```
 
-1. Set `KO_DATA_PATH=${GOPATH}/src/github.com/tektoncd/operator/cmd/operator/kodata`
+## Running Tests
 
-1. run `make ko-apply`
+[test docs](../test/README.md)
```
Comment (savitaashture) [inline docs/README.md:38]: As you have updated README can you please add instruction for cleanup also ?
`make clean-cluster`

**Review by savitaashture — COMMENTED** [review]
Thank you @nikhil-thomas for the README update :+1: 
PR looks good to me 

Just a minor suggestion
As you have updated README can you please add instruction for cleanup also ?
`make clean-cluster`

File: test/e2e/openshift/tektonaddondeployment_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	crNames := utils.ResourceNames{
+		TektonPipeline:  "pipeline",
+		TektonTrigger:   "trigger",
+		TektonAddon:     "addon",
+		Namespace:       "",
+		TargetNamespace: "tekton-pipelines",
+	}
+
+	utils.CleanupOnInterrupt(func() { utils.TearDownPipeline(clients, crNames.TektonPipeline) })
+	defer utils.TearDownPipeline(clients, crNames.TektonPipeline)
```
Comment (savitaashture) [inline test/e2e/openshift/tektonaddondeployment_test.go:44]: Can we add TearDown for triggers also as we are creating `triigers` here

File: test/resources/tektonaddons.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	mfc "github.com/manifestival/client-go-client"
+	mf "github.com/manifestival/manifestival"
+
+	"knative.dev/pkg/test/logging"
+
+	"github.com/tektoncd/operator/test/utils"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	pipelinev1alpha1 "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
```
Comment (savitaashture) [inline test/resources/tektonaddons.go:36]:
```suggestion
	addonv1alpha1 "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
```

File: test/resources/tektonaddons.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EnsureTektonAddonExists creates a TektonAddon with the name names.TektonAddon, if it does not exist.
+func EnsureTektonAddonExists(clients pipelinev1alpha1.TektonAddonInterface, names utils.ResourceNames) (*v1alpha1.TektonAddon, error) {
```
Comment (savitaashture) [inline test/resources/tektonaddons.go:42]:
```suggestion
func EnsureTektonAddonExists(clients addonv1alpha1.TektonAddonInterface, names utils.ResourceNames) (*v1alpha1.TektonAddon, error) {
```

File: test/resources/tektonaddons.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		return clients.Create(context.TODO(), ks, metav1.CreateOptions{})
+	}
+	return ks, err
+}
+
+// WaitForTektonAddonState polls the status of the TektonAddon called name
+// from client every `interval` until `inState` returns `true` indicating it
+// is done, returns an error or timeout.
+func WaitForTektonAddonState(clients pipelinev1alpha1.TektonAddonInterface, name string,
```
Comment (savitaashture) [inline test/resources/tektonaddons.go:64]:
```suggestion
func WaitForTektonAddonState(clients addonv1alpha1.TektonAddonInterface, name string,
```

File: test/resources/tektonaddons.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	defer span.End()
+
+	var lastState *v1alpha1.TektonAddon
+	waitErr := wait.PollImmediate(Interval, Timeout, func() (bool, error) {
+		lastState, err := clients.Get(context.TODO(), name, metav1.GetOptions{})
+		return inState(lastState, err)
+	})
+
+	if waitErr != nil {
+		return lastState, fmt.Errorf("tektonpipeline %s is not in desired state, got: %+v: %w", name, lastState, waitErr)
```
Comment (savitaashture) [inline test/resources/tektonaddons.go:76]:
```suggestion
		return lastState, fmt.Errorf("tektonaddon %s is not in desired state, got: %+v: %w", name, lastState, waitErr)
```

File: test/resources/tektonaddons.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func verifyNoTektonAddonCR(clients *utils.Clients) error {
+	pipelines, err := clients.TektonAddonAll().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(pipelines.Items) > 0 {
+		return errors.New("Unable to verify cluster-scoped resources are deleted if any TektonAddon exists")
+	}
```
Comment (savitaashture) [inline test/resources/tektonaddons.go:139]:
```suggestion
	addons, err := clients.TektonAddonAll().List(context.TODO(), metav1.ListOptions{})
	if err != nil {
		return err
	}
	if len(addons.Items) > 0 {
		return errors.New("Unable to verify cluster-scoped resources are deleted if any TektonAddon exists")
	}
```

**Review by savitaashture — COMMENTED** [review]
few nits

File: test/e2e/openshift/tektonaddondeployment_test.go (hunk 6 above)
Comment (nikhil-thomas) [reply test/e2e/openshift/tektonaddondeployment_test.go:44]: :heavy_check_mark:

File: test/resources/tektonaddons.go (hunk 7 above)
Comment (nikhil-thomas) [reply test/resources/tektonaddons.go:36]: :heavy_check_mark:

File: test/e2e/openshift/tektonaddondeployment_test.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+	utils.CleanupOnInterrupt(func() { utils.TearDownTrigger(clients, crNames.TektonPipeline) })
```
Comment (savitaashture) [inline test/e2e/openshift/tektonaddondeployment_test.go:46]:
```suggestion
	utils.CleanupOnInterrupt(func() { utils.TearDownTrigger(clients, crNames.TektonTrigger) })
```

File: test/e2e/openshift/tektonaddondeployment_test.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	defer utils.TearDownTrigger(clients, crNames.TektonTrigger)
+
+	utils.CleanupOnInterrupt(func() { utils.TearDownAddon(clients, crNames.TektonPipeline) })
```
Comment (savitaashture) [inline test/e2e/openshift/tektonaddondeployment_test.go:49]:
```suggestion
	utils.CleanupOnInterrupt(func() { utils.TearDownAddon(clients, crNames.TektonAddon) })
```

**Pull request merged**